src/kernel/arch/aarch64/boot.zig

```diff
@@ -1,49 +1,57 @@
-const rpi = @import("rpi.zig");
 const arch = @import("arch.zig");
+const cpu = arch.cpu;
 const interrupts = @import("interrupts.zig");
+const rpi = @import("rpi.zig");
 
-export var kernel_stack: [16 * 1024]u8 align(16) linksection(".bss.stack") = undefined;
+const number_of_cores = 4;
+const per_core_stack_size = 16 * 1024;
+export var kernel_stack: [number_of_cores * per_core_stack_size]u8 align(16) linksection(".bss.stack") = undefined;
 
 extern fn kmain(payload: *const rpi.RaspberryPiBoard) void;
 extern var KERNEL_STACK_END: *usize;
 
 export fn _start() linksection(".text.boot") callconv(.Naked) noreturn {
-    // Halt all cores other than the primary core, until the kernel has multicore support
-    // This must be in assembly and not zig due to the fact the zig compiled code
-    //  requires the sp and the sp is not initialized (and cannot be initialized) at this point.,
+    // Give each of four cores one-fourth of the reserved stack space
     asm volatile (
         \\ mrs x0, mpidr_el1
-        \\ mov x1, #3
-        \\ ands x0, x0, x1
-        \\ beq cpu0
-        \\hang:
-        \\ wfe
-        \\ b hang
-        \\cpu0:
-    );
-
-    // Setup the stack
-    asm volatile (
-        \\mov sp, %[stack_end]
+        \\ and x0, x0, #0x3
+        \\ add x0, x0, #1     // core number + 1 therefore 1..4
+        \\ mov x1, #16 * 1024 // per_core_stack_size
+        \\ mul x0, x0, x1
+        \\ ldr x1, =kernel_stack
+        \\ add x0, x0, x1
+        \\ mov sp, x0
+        :
         :
-        : [stack_end] "r" (@ptrCast([*]u8, &KERNEL_STACK_END))
+        : "x0", "x1"
     );
 
-    interrupts.exception_handler_depth = 0;
+    start(); // must start a proper zig function to get x29 (frame pointer) initialized!
+}
 
-    // Setup the exception table
-    asm volatile (
-        \\ msr vbar_el1, %[table_addr]
-        \\ msr vbar_el2, %[table_addr]
-        \\ msr vbar_el3, %[table_addr]
-        :
-        : [table_addr] "r" (@ptrToInt(&interrupts.exception_table))
-    );
+fn start() noreturn {
+    // Halt all cores other than the core 0, until the kernel has multicore support
+    if (cpu.mpidr.el(1).read() & 0x3 != 0) {
+        while (true) {
+            cpu.wfe();
+        }
+    }
+
+    // Give all exception levels the same vector table
+    cpu.vbar.el(1).write(@ptrToInt(&interrupts.exception_table));
+    cpu.vbar.el(2).write(@ptrToInt(&interrupts.exception_table));
+    cpu.vbar.el(3).write(@ptrToInt(&interrupts.exception_table));
+
+    interrupts.exception_handler_depth = 0;
 
     arch.enableFlatMmu();
 
-    // The rpi puts the board part number in midr_el1
     board = detectBoard();
+    arch.initMmioAddress(&board);
+
+    // invoke alignment exception
+    // var addr: usize = 1;
+    // @intToPtr(*volatile u32, addr).* = 0;
```

Comment:
This looks like it's for testing so remember to remove it before we merge.

Reply:
Done.